### PR TITLE
Remote training using TrainConfig and torch.jit

### LINKED ISSUE #2162: POC: torchscript serializable along with syft operations
In order to implement plans with torchscript we need a proof of concept showing that is possible to serialize torchscript along with syft operations.

Per instance:

```python
import torch
import syft
hook = syft.TorchHook(torch)

def foo(x, y):
    return x + y

foo(torch.rand(3), torch.rand(3))
traced_foo = torch.jit.trace(foo, (torch.rand(3), torch.rand(3)))

# Should return a fixed precision tensor as well
traced_foo(torch.rand(3).fix_prec(), torch_rand(3).fix_prec())
```

Should ideally work out of the box.

### PR DESCRIPTION
**Pull request by midokura-silvia**
This pull request enables a remote training of a model without intermediate communication between the server and the remote worker. 

Ingredients:
- generic pointers and object wrappers, allowing to point to objects other than tensors (e.g. torch.jit.TopLevelTracedModule or functions)
- TrainConfig and FederatedClient implementations of @mari-linhares 
- serialization of pytorch models via torch.jit
- working example how to trace a model, send it to a websocket worker and train it on the available local data without intermediate communication.

Fixes #2162

### CODE REVIEW

**Comment by midokura-silvia** [comment]
Travis is failing due to a problem with torch 1.1.0, unit tests run through with torch 1.0.1.
See pytorch issue 20017, a fix was merged in pytorch via pull request 20386

File: examples/experimental/Federated Learning with TrainConfig/run_training_train_config.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import torch as th
```
Comment (mari-linhares) [inline examples/experimental/Federated Learning with TrainConfig/run_training_train_config.py]: Delete this file. We can keep the notebook version only.

File: examples/experimental/Federated Learning with TrainConfig/Federated learning with TrainConfig.ipynb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    "        x = F.relu(self.fc2(x))\n",
+    "        x = self.fc3(x)\n",
+    "        return x\n",
+    "\n",
+    "model = Net()\n",
+    "data = th.tensor(th.tensor([[-10, -2.0], [1, 1.1], [11, 22.1], [-10, 1.2]]))\n",
+    "\n",
+    "traced_model = th.jit.trace(model, data)\n",
+    "\n",
+    "model_with_id = pointers.ObjectWrapper(id=sy.ID_PROVIDER.pop(), obj=traced_model)\n",
```
Comment (mari-linhares) [inline examples/experimental/Federated Learning with TrainConfig/Federated learning with TrainConfig.ipynb]:
```
traced_model = th.jit.trace(model, data)

model_with_id = pointers.ObjectWrapper(id=sy.ID_PROVIDER.pop(), obj=traced_model)
loss_fn_with_id = pointers.ObjectWrapper(id=sy.ID_PROVIDER.pop(), obj=loss_fn)

model_ptr = me.send(model_with_id, alice)
loss_fn_ptr = me.send(loss_fn_with_id, alice)
```

This should be done internally by trainconfig.

File: examples/experimental/Federated Learning with TrainConfig/Federated learning with TrainConfig.ipynb [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      "Iteration 4: alice's loss: tensor(0.2458, requires_grad=True)\n",
+      "\n",
+      "Evaluation after training:\n",
+      "Loss: 0.24460731446743011\n"
+     ]
+    }
+   ],
+   "source": [
+    "for epoch in range(5):\n",
+    "    loss = alice.fit(dataset_key=\"vectors\", return_id=88 + epoch)\n",
```
Comment (mari-linhares) [inline examples/experimental/Federated Learning with TrainConfig/Federated learning with TrainConfig.ipynb]: return_id could be a string.

File: examples/experimental/Federated Learning with TrainConfig/run_websocket_server.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+def start_proc(participant, kwargs):  # pragma: no cover
+    """ helper function for spinning up a websocket participant """
+
+    def target():
+        server = participant(**kwargs)
+
+        # data = AutogradTensor().on(th.tensor([[-1, 2.0], [0, 1.1], [-1, 2.1], [0, 1.2]], requires_grad=True))
+        # target = AutogradTensor().on(th.tensor([[1.0], [0.0], [1.0], [0.0]], requires_grad=True))
+        data = th.tensor([[-1, 2.0], [0, 1.1], [-1, 2.1], [0, 1.2]], requires_grad=True)
```
Comment (mari-linhares) [inline examples/experimental/Federated Learning with TrainConfig/run_websocket_server.py]: We need to:

1. Clean up this file
2. Instantiate multiple remote workers
3. Split some real dataset (MNIST probably) between users

File: syft/federated/federated_client.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            self.optimizer = th.optim.SGD(model.parameters(), lr=lr)
+        else:
+            raise ValueError("Unknown optimizer: {}".format(optimizer_name))
+        return self.optimizer
+
+    def fit_batch(self, data, target, *args, **kwargs):
+        if self.verbose:
+            logger.debug("data = %s", data)
+            logger.debug("target = %s", target)
+        self.register_obj(data)
```
Comment (mari-linhares) [inline syft/federated/federated_client.py]: I don't like that we need to register/deregister tensors explicitly.

File: syft/federated/federated_client.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            sampler = SequentialSampler(data_range)
+
+        batch_sampler = BatchSampler(sampler, self.train_config.batch_size, drop_last)
+        return batch_sampler
+
+    def _fit(self, model, key):
+        loss_fn = self.get_obj(self.train_config.loss_plan_id).obj
+        model.train()
+        logger.setLevel(logging.DEBUG)
```
Comment (mari-linhares) [inline syft/federated/federated_client.py]: remove logger

File: test/federated/test_train_config.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+    print("Evaluation before training")
+    pred = model(data)
+    loss_before = loss_fn(real=target, pred=pred)
+    print("Loss: {}".format(loss_before))
+
+    print("Evaluation after training:")
+    pred = new_model(data)
+    loss_after = loss_fn(real=target, pred=pred)
+    print("Loss: {}".format(loss_after))
```
Comment (mari-linhares) [inline test/federated/test_train_config.py]: remove print statements

File: syft/federated/federated_client.py (hunk 5 above)
Comment (midokura-silvia) [reply syft/federated/federated_client.py]: The fit_batch method still relies on the previous train_config content with forward_plans etc.. I will update it. The call to register the data will go away. It was needed for plans to work.

File: .travis.yml [hunk 8]
```diff
@@ -39,7 +39,8 @@ script:
   - coverage run --omit=$OMIT_FROM_COVERAGE -m pytest
   - coverage report --fail-under 95 -m
   # ensuring coverage of tests stays at 100 until entire coverage can be 100
-  - coverage report --fail-under 100 -m --include="test/*"
+  # reduced to 99 to allow skipping of some tests
+  - coverage report --fail-under 99 -m --include="test/*"
```
[the remark this answers is not in the document]
Comment (midokura-silvia) [reply .travis.yml]: You're right. I corrected it already in the PR for serialization of jit script modules. The present PR sits on top of the other PR. So I just rebased and this modification went away.

File: examples/experimental/Federated Learning with TrainConfig/run_websocket_server.py (hunk 4 above)
Comment (midokura-silvia) [reply examples/experimental/Federated Learning with TrainConfig/run_websocket_server.py]: This is contained in #2217

**Comment by midokura-silvia** [comment]
Development moved to branch train-config-rebase

**Pull request closed without merging**